### DISCUSSION TITLE
Communication between VSCode and Theia extensions on backend

### DISCUSSION THREAD

**Discussion by xcariba** [eclipse-theia/theia#9370] [Q&A] [answered]
Hello!
I have a VSCode extension and Theia extension that implements some features that cannot be implemented as VSCode extension. I need to pass some object instance from VSCode extension to Theia extension. I found out that the only way to communicate between different types of extensions is commands. (as mentioned in #8126), so I created a command that returns this instance, but I cannot figure out how to execute command on backend.
I tried to inject CommandService or CommandRegistry/CommandRegistryImpl, but they are not binded on server side, but CommandService is binded on backend as frontend module (https://github.com/eclipse-theia/theia/blob/master/packages/core/src/node/backend-application-module.ts#L37).
I found that it is used in debugger backend, but it is passed in assistedInject method call and I cannot figure out how to get this instance.
Is there any way to execute command on backend in my extension or is there any other way to do so?

**Comment by kittaakos** [comment, marked as answer]
> communicate between different types of extensions is commands.

Yes, this is correct.

> Is there any way to execute command on backend

Why do you want to execute it on the backend? Can you share your code? Do you want to run some strickly Node stuff? If so, I would still do the following:
 - invoke the command on the frontend to get the _thing_ you need from the VS Code extension,
 - call your Theia backend service with the _thing_.

**Reply by kittaakos** [reply]
@xcariba, you can call any VS Code contributed command directly from the frontend. You have to take care of one thing, the VS Code extensions load lazily, so you need something like this in your code to make sure the commands are available:

```ts
@injectable()
class OneOfYourFrontendClasses {
  @inject(CommandService) cs: CommandService
  @inject(HostedPluginSupport) hps: HostedPluginSupport
  async foo(): Promise<any> {
    await this.hps.didStart
    return this.cs.executeCommand('yourCommandId')
  }
}
```

**Reply by xcariba** [reply]
Thank you for a reply!

Unfortunately I cannot share code. I want to pass something (some object) from vscode to extension to communicate with process (something like LSP) using rpc via stdio. I`m not sure if I can pass it from frontend to backend.

I looked into theia sources and found that, for example, debugger uses instance of CommandService to execute commands, I wanted to do something similar, but I don`t understand if it is possible to get CommenService instance on backend.

**Reply by kittaakos** [reply]
> but I don`t understand if it is possible to get CommenService instance on backend.

I cannot help much without understanding what you're doing, but I created a working example where the `CommandService` is injected into the backend: https://github.com/kittaakos/theia/commit/cbf197009501de364e31a1f8bcb18229c65d8e5a

Start the backend, connect to the backend from two browsers, run the `Execute example - Call CommandService on the backend` command from one browser, then from the other. See the notification. The trick is, even if the `CommadnService` is injected into the backend, the logic will be executed on the frontend. That means each connected Theia frontend will have a dedicated `SomeServiceImpl` instance on the backend.

You have to use connection scoped dependency injection.
 - https://community.theia-ide.org/t/connection-scoped-injection/224
 - https://community.theia-ide.org/t/multiple-clients-for-a-backend-services-json-rpc/860

I hope this helps.

**Reply by xcariba** [reply]
Thank you for an example, I was able to execute command located in my vscode extension, it works well for simple objects, but unfortunately it does not work with an interface implementation.
I'm trying to pass some kind of wrapper (interface implementation) from vscode extension to theia extension, so that my theia extension could call some method that is implemented in vscode extension and vice versa. I tried to return instance in vscode command, but I got `null` instead of a value (and it works well if I return some string, for example). 
It probably could be implemented with command calls in both extensions, but I don't think that it is a good solution, so I want to do it more... "natively".

**Reply by xcariba** [reply]
I finally implemented communication between vscode and theia extensions by registering commands in both extensions and executing them from both sides when needed (something like rpc based on commands). I dont think it is the best solution, but it is probably the only way for now. Thank you for your help!
